docs: 5 type declarations (#5159)
dfinity/motoko: doc/md/fundamentals/3-declarations/5-type-declarations.md

```

## Parameterized (generic) types

Type declarations can be parameterized, allowing them to work flexibly with multiple types while ensuring type safety. This lets you create generic and reusable type definitions that adapt to different data types as needed.

```motoko no-repl
// `Box<T> is a generic type where T represents any type.
type Box<T> = {
    value: T;
};

// numberBox stores a Nat and textBox stores a Text.
let numberBox : Box<Nat> = { value = 42 };
let textBox : Box<Text> = { value = "Hello" };
```

## Recursive types and productivity

Motoko allows recursive type definitions as long as they are productive. This means that any recursion in a type must pass through a constructor (such as an option (`?`), a variant, or a record field) before referring back to itself.

#### Productive recursive type example

```motoko no-repl
type List<T> = ?(T, List<T>);
```

`List<T>` defines a linked list where each node holds a value of type `T` and points to either another `List<T>` or `null` to mark the end. Since the recursion passes through a constructor (`?`), this type is productive and accepted by the compiler.

#### Non-productive recursive type example

```motoko no-repl

> type C = C; // This definition infinitely refers to itself

Reviewer: ```suggestion
type C = C; // This definition immediately refers to itself
```

```

> This type never resolves to a concrete value because it lacks a constructor in its recursion; it is non-productive and rejected by the compiler.

Reviewer: ```suggestion
This type never produces a concrete type when expanded because its definition is too cyclic; it is reported as a _non-productive_  type definition and rejected by the compiler.
```

## Expansiveness in type definitions

Motoko enforces non-expansiveness to ensure type definitions do not expand indefinitely, keeping types well-formed and preventing infinite or unmanageable type structures.

#### Non-expansive

```motoko no-repl

> /// Expands without introducing a larger type.

Reviewer: ```suggestion
// Expands without introducing an infinite set of new instances of `List<_>`.
```
